cpp/log/cluster_state_controller-inl.h

```diff
@@ -0,0 +1,203 @@
+ClusterStateController<Logged>::ClusterStateController(
+    util::Executor* executor, ConsistentStore<Logged>* store,
+    const MasterElection* election, const int min_serving_nodes,
+    const double min_serving_fraction)
+    : executor_(CHECK_NOTNULL(executor)),
+      store_(CHECK_NOTNULL(store)),
+      election_(CHECK_NOTNULL(election)),
+      min_serving_nodes_(min_serving_nodes),
+      min_serving_fraction_(min_serving_fraction),
+      exiting_(false),
+      update_required_(false),
+      cluster_serving_sth_update_thread_(
+          std::bind(&ClusterStateController<Logged>::ClusterServingSTHUpdater,
+                    this)) {
+  store_->WatchClusterNodeStates(
+      std::bind(&ClusterStateController::OnClusterStateUpdated, this,
+                std::placeholders::_1));
+}
+
+
+template <class Logged>
+ClusterStateController<Logged>::~ClusterStateController() {
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    exiting_ = true;
+    update_required_cv_.notify_all();
+  }
+  cluster_serving_sth_update_thread_.join();
+}
+
+
+template <class Logged>
+void ClusterStateController<Logged>::NewTreeHead(
+    const ct::SignedTreeHead& sth) {
+  CHECK(sth.IsInitialized());
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (local_node_state_.has_newest_sth()) {
+    CHECK_GE(sth.timestamp(), local_node_state_.newest_sth().timestamp());
+  }
+  local_node_state_.mutable_newest_sth()->CopyFrom(sth);
+  PushLocalNodeState(lock);
+}
+
+
+template <class Logged>
+void ClusterStateController<Logged>::ContiguousTreeSizeUpdated(
+    const int64_t new_contiguous_tree_size) {
+  CHECK_GE(new_contiguous_tree_size, 0);
+  std::unique_lock<std::mutex> lock(mutex_);
+  CHECK_GE(new_contiguous_tree_size, local_node_state_.contiguous_tree_size());
+  local_node_state_.set_contiguous_tree_size(new_contiguous_tree_size);
+  PushLocalNodeState(lock);
+}
+
+
+template <class Logged>
+bool ClusterStateController<Logged>::GetCalculatedServingSTH(
+    SignedTreeHead* sth) const {
+  CHECK_NOTNULL(sth);
+  std::lock_guard<std::mutex> lock(mutex_);
+  if (calculated_serving_sth_.get() == nullptr) {
+    return false;
+  }
+  *sth = *calculated_serving_sth_;
+  return true;
+}
+
+
+template <class Logged>
+void ClusterStateController<Logged>::PushLocalNodeState(
+    const std::unique_lock<std::mutex>& lock) {
+  CHECK(lock.owns_lock());
+  const util::Status status(store_->SetClusterNodeState(local_node_state_));
+  if (!status.ok()) {
+    LOG(WARNING) << status;
+  }
+}
+
+
+template <class Logged>
+void ClusterStateController<Logged>::OnClusterStateUpdated(
+    const std::vector<Update<ct::ClusterNodeState>>& updates) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  for (const auto& update : updates) {
+    const std::string& node_id(update.handle_.Entry().node_id());
+    if (update.exists_) {
+      all_node_states_[node_id] = update.handle_.Entry();
+    } else {
+      CHECK_EQ(1, all_node_states_.erase(node_id));
```

Comment:
So if we lose nodes to the point where nobody has enough entries to serve the current STH, we just won't update the serving STH (```CalculateServingSTH()``` below will return false), and indirectly stop serving (as all the nodes left would be lame ducked), I suppose?